### PR TITLE
the scheduling queue logs the error and treats it as QueueAfterBackoff

### PR DESCRIPTION
**Pull request by carlory**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

See https://github.com/kubernetes/kubernetes/pull/119155#discussion_r1262217913

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
QueueingHint got error in its returning value. If QueueingHint returns error, the scheduler logs the error and treats the event as QueueAfterBackoff so that the Pod wouldn't be stuck in the unschedulable pod pool.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:295]: Could you remove all error logs before returning error on the plugin side and put useful information into returning error instead?

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:331]: need to create a correct `err`

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue.go:396]:
```suggestion
				logger.Error(err, "QueueingHintFn returns error", "plugin", hintfn.PluginName, "pod", klog.KObj(pod))
```

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue.go:395]: Please have the same comment in the QueueingHintFn comment (pkg/scheduler/framework/types.go).

File: pkg/scheduler/internal/queue/scheduling_queue_test.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:2419]:
```suggestion
			name: "Treat the event as QueueAfterBackoff when QueueHintFn returns error",
```

File: pkg/scheduler/internal/queue/scheduling_queue_test.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:2379]:
```suggestion
	queueHintReturnErr := func(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
```

**Review by sanposhiho — COMMENTED** [review]
Thanks for a prompt action!

**Comment by carlory** [comment]
@sanposhiho updated.

File: pkg/scheduler/framework/types.go
Comment (sanposhiho) [inline pkg/scheduler/framework/types.go:95]:
```suggestion
// It's called before a Pod gets moved from unschedulableQ to backoffQ or activeQ.
// If it returns an error, the scheduling queue behaves like QueueingHintFn returned QueueAfterBackoff so that
// we can prevent the Pod from stucking in the unschedulable pod pool.
```

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue.go:395]:
```suggestion
				// the Pod from stucking in the unschedulable pod pool.
```

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue.go:396]: Let's include oldObj and newObj in the log

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue.go:394]:
```suggestion
				// If the QueueingHintFn returned an error, we should treat the event as QueueAfterBackoff so that we can prevent
```

**Comment by carlory** [comment]
@sanposhiho I pushed some changes. please review again.

File: pkg/scheduler/framework/types.go
Comment (codefromthecrypt) [inline pkg/scheduler/framework/types.go:102]: since breaking api anyway, we can add ctx as a parameter, since almost everything in the framework accepts one (except this and one other place I noticed)

Comment (carlory) [reply pkg/scheduler/framework/types.go:102]: cc @sanposhiho what do you think?

There's some discuss about logger and ctx. https://github.com/kubernetes/kubernetes/pull/119155#discussion_r1260129257

Comment (sanposhiho) [reply pkg/scheduler/framework/types.go:102]: Ah, maybe Aldo missed my followup question on ctx in the original thread. Let me follow up again in the original PR.

I'd like to make this PR's scope only change for error. We can easily have another change for ctx if we conclude we have.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:981]: small nit

```suggestion
				return
			}
				
			require.NoError(t, err)
			require.Equal(t, tc.expectedHint, actualHint)
			
```

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:1106]: ditto

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue.go:396]: (klog.KObj() can handle nil)


```suggestion
				logger.Error(err, "QueueingHintFn returns error", "plugin", hintfn.PluginName, "pod", klog.KObj(pod), "oldObj", klog.KObj(oldObj), "newObj", klog.KObj(newObj))
```

**Review by sanposhiho — COMMENTED** [review]
tiny nits only. will /lgtm after them

/approve

File: pkg/scheduler/framework/types.go
Comment (sanposhiho) [reply pkg/scheduler/framework/types.go:102]: https://github.com/kubernetes/kubernetes/pull/119155/files#r1263484074

**Comment by carlory** [comment]
@sanposhiho I pushed some changes, please review again. Thanks!

File: pkg/scheduler/scheduler_test.go
Comment (kerthcet) [inline pkg/scheduler/scheduler_test.go:812]: Can you provide one more case with error returned?

File: pkg/scheduler/framework/types.go
Comment (kerthcet) [inline pkg/scheduler/framework/types.go:94]: Let's make it more precise, like
```suggestion
// If it returns an error, we'll take the returned QueueingHint as `QueueAfterBackoff` at the caller whatever we returned here so that
```

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue.go:396]: @carlory 

Ah, so we cannot use interface{} in `klog.KObj`...
Cannot we somehow utilize As function here to convert two objects more wisely?

Comment (carlory) [reply pkg/scheduler/internal/queue/scheduling_queue.go:396]: @sanposhiho 

```go
func As[T runtime.Object](oldObj, newobj interface{}) (T, T, error)
```

I'm trying to use this function to convert the oldObj and newObj to the klog.KMetadata type. But it doesn't work.

The As function only convert the oldObj and newObj to the runtime.Object type. And I cannot find any existed function to convert the interface{} to the klog.KMetadata type.

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (carlory) [reply pkg/scheduler/internal/queue/scheduling_queue.go:398]: cc @sanposhiho should I log it?

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (carlory) [reply pkg/scheduler/internal/queue/scheduling_queue.go:396]: After change the func to 

```go
func As[T any](oldObj, newobj interface{}) (T, T, error)
```

It works.

Comment (carlory) [reply pkg/scheduler/internal/queue/scheduling_queue.go:396]: After change the func to 

```go
func As[T any](oldObj, newobj interface{}) (T, T, error)
```

It works.

File: pkg/scheduler/scheduler_test.go
Comment (carlory) [reply pkg/scheduler/scheduler_test.go:914]: @kerthcet added.

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: I'd like to keep log with oldObj, newObj

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue.go:433]: Can we handle err from util.As just in case?

**Review by sanposhiho — COMMENTED** [review]
/lgtm

Leave @kerthcet for approval in case he has something more to say.

**Comment by sanposhiho** [comment]
/remove-approve
/unhold

To keep it to @kerthcet https://github.com/kubernetes/kubernetes/pull/119290#pullrequestreview-1536782429

**Comment by sanposhiho** [comment]
Sorry for the mess, seems just takes longer to remove the approval

/unhold

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (wackxu) [inline pkg/scheduler/internal/queue/scheduling_queue.go:433]:
```suggestion
				oldObjMeta, newObjMeta, asErr := util.As[klog.KMetadata](oldObj, newObj)
```

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (wackxu) [inline pkg/scheduler/internal/queue/scheduling_queue.go:434]: also log the asErr

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue.go:435]: Given this error shouldn't happen + too internal for users, it feels like we don't need this error log. Just a below log is enough.

File: pkg/scheduler/scheduler_test.go
Comment (sanposhiho) [reply pkg/scheduler/scheduler_test.go:817]: The reason we're comparing the result like this is Golang cannot compare the function is the same. 
So, we change the result coming from QueueingHintFn in every fake plugin and check registered functions are expected ones by checking the result from QueueingHint

```
if fn.QueueingHintFn(logger, nil, nil, nil) != wantfns[i].QueueingHintFn(logger, nil, nil, nil) {
```

File: pkg/scheduler/util/utils.go
Comment (sanposhiho) [inline pkg/scheduler/util/utils.go:169]: Let's add one unit test case to confirm we can convert objects to klog.KMetadata

File: pkg/scheduler/scheduler_test.go
Comment (kerthcet) [reply pkg/scheduler/scheduler_test.go]: Then no need to compare `QueueingHintFn` also, because `pluginName` tells everything.

File: pkg/scheduler/util/utils.go
Comment (carlory) [reply pkg/scheduler/util/utils.go:169]: @sanposhiho added.

**Review by sanposhiho — COMMENTED** [review]
Perfect, thanks for a long journey here!

/lgtm
/hold

Please add the release note. (someone may already be using the queueing hint in out-of-tree plugins) Once done, I'll `/approve`.

**Comment by sanposhiho** [comment]
/release-note-edit 
```release-note 
QueueingHint got error in its returning value. If QueueingHint returns error, the scheduler logs the error and treats the event as QueueAfterBackoff so that the Pod wouldn't be stuck in the unschedulable pod pool.
```

**Review by sanposhiho — APPROVED** [review]
It works.

/approve

**Pull request merged**
